benchmark/main.cpp

```diff
@@ -56,979 +50,1350 @@
+#if defined(MBEDTLS_AES_C) && defined(MBEDTLS_GCM_C)
+MBED_NOINLINE static int benchmark_aes_gcm()
+{
+    int ret = 0;
+    int keysize;
+    mbedtls_gcm_context gcm;
+
+    mbedtls_gcm_init(&gcm);
+
+    for (keysize = 128; keysize <= 256; keysize += 64) {
+        ret = mbedtls_snprintf(title, sizeof(title), "AES-GCM-%d", keysize);
+        if (ret < 0 || static_cast<size_t>(ret) >= sizeof(title)) {
+            mbedtls_printf("Failed to compose title string using "
+                           "mbedtls_snprintf(): %d\n", ret);
+            goto exit;
+        }
 
-            TIME_AND_TSC( title,
-                    mbedtls_gcm_crypt_and_tag( &gcm, MBEDTLS_GCM_ENCRYPT, BUFSIZE, tmp,
-                        12, NULL, 0, buf, buf, 16, tmp ) );
+        memset(buf, 0, sizeof(buf));
+        memset(tmp, 0, sizeof(tmp));
 
-            mbedtls_gcm_free( &gcm );
+        ret = mbedtls_gcm_setkey(&gcm, MBEDTLS_CIPHER_ID_AES, tmp, keysize);
+        if (ret == MBEDTLS_ERR_AES_FEATURE_UNAVAILABLE) {
+            continue;
+        } else if (ret != 0) {
+            PRINT_ERROR(ret, "mbedtls_gcm_setkey()");
+            goto exit;
         }
+
+        BENCHMARK_FUNC_CALL(title,
+                            mbedtls_gcm_crypt_and_tag(&gcm,
+                                    MBEDTLS_GCM_ENCRYPT,
+                                    BUFSIZE, tmp, 12, NULL,
+                                    0, buf, buf, 16, tmp));
     }
-#endif
-#if defined(MBEDTLS_CCM_C)
-    if( todo->aes_ccm )
-    {
-        int keysize;
-        mbedtls_ccm_context ccm;
-
-        mbedtls_ccm_init( &ccm );
-        for( keysize = 128; keysize <= 256; keysize += 64 )
-        {
-            mbedtls_snprintf( title, sizeof( title ), "AES-CCM-%d", keysize );
-
-            memset( buf, 0, sizeof( buf ) );
-            memset( tmp, 0, sizeof( tmp ) );
-            mbedtls_ccm_setkey( &ccm, MBEDTLS_CIPHER_ID_AES, tmp, keysize );
-
-            TIME_AND_TSC( title,
-                    mbedtls_ccm_encrypt_and_tag( &ccm, BUFSIZE, tmp,
-                        12, NULL, 0, buf, buf, tmp, 16 ) );
-
-            mbedtls_ccm_free( &ccm );
+
+exit:
+    mbedtls_gcm_free(&gcm);
+
+    return ret;
+}
+#endif /* MBEDTLS_AES_C && MBEDTLS_GCM_C */
+
+#if defined(MBEDTLS_AES_C) && defined(MBEDTLS_CCM_C)
+MBED_NOINLINE static int benchmark_aes_ccm()
+{
+    int ret = 0;
+    int keysize;
+    mbedtls_ccm_context ccm;
+
+    mbedtls_ccm_init(&ccm);
+
+    for (keysize = 128; keysize <= 256; keysize += 64) {
+        ret = mbedtls_snprintf(title, sizeof(title), "AES-CCM-%d", keysize);
+        if (ret < 0 || static_cast<size_t>(ret) >= sizeof(title)) {
+            mbedtls_printf("Failed to compose title string using "
+                           "mbedtls_snprintf(): %d\n", ret);
+            goto exit;
+        }
+
+        memset(buf, 0, sizeof(buf));
+        memset(tmp, 0, sizeof(tmp));
+
+        ret = mbedtls_ccm_setkey(&ccm, MBEDTLS_CIPHER_ID_AES, tmp, keysize);
+        if (ret != 0) {
```

Comment:
CCM could also return `MBEDTLS_AES_FEATURE_NOT_SUPPORTED` in case no alternative CCM